prime-router/docs/design/design/auth/auth-design.md

```diff
@@ -0,0 +1,347 @@
+
+### Spring reverse proxy microservice
+
+This approach would create a new microservice whose sole responsibility is authn/authz. It would sit in front of
+all requests to protected endpoints.
+
+### Prerequisites
+- Designate protected endpoints and put them behind a VPC.
+  - Traffic must flow through our authn/authz service  
+  - Protected endpoints should not be accessible to the public internet
+
+A typical report request would look like this
+- Client makes request to authorization server (or us in the ReportStream implementation solution) to retrieve token with their own credentials
+- Client uses retrieved token with request to ResportStream
+- Request comes in and hits our auth microservice instead of ResportStream directly
+- Depending on our implementation, we decide if that request is legitimate and allowed to access the designated endpoint
+  - In cloud-based solutions, we would make calls out to an external authz server
+    - We could also introspect the token locally if speed is a factor but its more code to write/maintain
+  - In a ReportStream implementation solution, we would introspect certain cookies and headers and compare them to what we have in a datastore
+- If the request is unauthenticated or unauthorized, short-circuit the request to a 401
+- If the request is authenticated and authorized for the resource, pass it along to the internal protected endpoint
+
+Pros
+- Extremely scalable
+- All authn/authz code is in one place
+- Security concerns can be centralized as all requests go through a single public endpoint
+- Simplifies business logic as it no longer has to worry about authn/authz
+- Extensible outside of core RS services if required
+
+Cons
+- Highly complex setup from both an engineering and devops perspective
+- New service needs to be written and maintained
+- A single point of failure if the service goes down
+- Can lack customization options for very specific use-cases
+
+Code example:
+```kotlin
+
+// handles all incoming requests
+@RequestMapping("/**")
+fun protectedReadEndpoint(
+    @RequestBody(required = false) body: String,
+    method: HttpMethod, 
+    request: HttpServletRequest,
+    response: HttpServletResponse
+) {
+    // this is all psudeocode!
+    
+    // check if we need to auth at all
+    val authRequired = condigService.isAuthRequired(request.path, method)
+    
+    return if (authRequired) {
+        // call out to authn/authz service
+        val authnResult = authService.checkToken(request)
+        val authzResult = authzService.checkAuthorization()
+        
+        // possible results
+        when {
+            authnResult == Authenticated && authzResult == Authorized -> httpClient.request(method, protectedEndpoint, body)
+            else -> UnauthorizedResponse
+        }
+    } else {
+        // pass through if no auth needed
+        httpClient.request(method, protectedEndpoint, body)
+    }
+}
+```
+
+Authz configuration file example:
+```json
```

Comment:
Config options to match client id

Reply:
claim vs scope

Reply:
add potential option of auth service not as a proxy

Reply:
investigate if its just overkill to match up client id given our users are trusted

Reply:
Add the options for how senders request their access token

Reply:
options on visibility outside organization